packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
+
+        if (txConfirmed) {
+          this.setEventInfo("block", wrappedListener, {
+            hasBeenConfirmed: true,
+          });
+        } else {
+          const { hasBeenConfirmed } = this.getEventInfo(
+            "block",
+            wrappedListener
+          );
+          if (hasBeenConfirmed)
+            this.emitErrorEvent(
+              new Error(`Your message has been included in an uncle block.`)
+            );
+
+          return;
+        }
+
+        const { confirmations } = receipt;
+        const message = {
+          data: {
+            confirmations: confirmations,
+          },
+        };
+
+        await listener(message);
       } catch (error) {
-        throw new Error(`Callback function with error: ${error.message}`);
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
       }
-    });
+    };
+
+    this.addListener(eventName, "block", listener, wrappedListener);
   };
 
-  removeAllListeners = async () => {
+  // Tech debt
+  // This method avoid duplicated nonce generation when rapid succession of several transactions
+  // See: https://github.com/ethereumbook/ethereumbook/blob/04f66ae45cd9405cce04a088556144be11979699/06transactions.asciidoc#keeping-track-of-nonces
+  // How we'll should keeping track of nonces?
+  waitForMessage = async () => {
     const tx = await this.#txPromise;
-    this.#provider.removeAllListeners(tx.hash);
+
+    await this.#provider.waitForTransaction(tx.hash);
   };
 }
 
-class EventSubscription {
-  #eventNames;
+class ContractSubscription extends Subscription {
+  #contract;
   #provider;
+  #eventNames;
 
-  constructor(eventNames, provider) {
-    this.#eventNames = eventNames;
+  // Why eventNames shoud be passed here?
+  constructor(contract, provider, eventNames) {
+    eventNames.forEach(eventName => {
+      if (contract.interface.events[eventName] === undefined)
+        throw new Error(`Event '${eventName}' not found.`);
+    });
+
+    super(contract);
+    this.#contract = contract;
     this.#provider = provider;
+    this.#eventNames = eventNames;
   }
 
-  on = (eventName, callback) => {
-    if (!this.#eventNames.includes(eventName))
+  on = (eventName, listener) => {
+    if (!this.#eventNames.includes(eventName) && eventName !== "error") {
       throw new Error(
         `This subscription isn't subscribed on '${eventName}' event.`
       );
+    }
+
+    if (eventName === "error") {
+      this.addErrorListener(listener);
+      return;
+    }
+
+    const wrappedListener = async (...args) => {
+      const event = args.pop();
+      const message = {
+        data: {
+          args: event.args,
+        },
+      };
+
+      try {
+        await listener(message);
+      } catch (error) {
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
+      }
+    };
+
+    this.addListener(eventName, eventName, listener, wrappedListener);
```

Comment:
Refs: https://github.com/tasitlabs/TasitSDK/pull/87#discussion_r245971525

That way was the way that I've found to parse/tracking user/our listener. If we'll not expose `s .removeListener(eventName, listener)` I think that we can remove this. Otherwise, I do not see another way to do that for now.

Reply:
I think we will want to have a removeListener function, so we need some form of addListener function. We can decide on the exact params needed in that other thread you linked to. I'll resolve this one. Thanks!